File: Fort.ind UWP/Services/ProfileService.vb
Imports System.Security.Cryptography
Imports System.Text

''' <summary>
''' Service for managing user accounts and authentication
''' Handles login, registration, and profile management
''' Ready to be extended with server authentication later
''' </summary>
Public Class ProfileService

    ''' <summary>
    ''' The currently logged-in user profile
    ''' </summary>
    Public Shared Property CurrentUser As UserProfile

    ''' <summary>
    ''' Event raised when user logs in or out
    ''' </summary>
    Public Shared Event AuthStateChanged As EventHandler(Of Boolean)

    ''' <summary>
    ''' Attempts to register a new user
    ''' </summary>
    Public Shared Async Function RegisterAsync(username As String, password As String, displayName As String, email As String) As Task(Of RegistrationResult)
        ' Validate inputs
        If String.IsNullOrWhiteSpace(username) OrElse username.Length < 3 Then
            Return New RegistrationResult(False, "Username must be at least 3 characters")
        End If

        If String.IsNullOrWhiteSpace(password) OrElse password.Length < 4 Then
            Return New RegistrationResult(False, "Password must be at least 4 characters")
        End If

        If String.IsNullOrWhiteSpace(displayName) Then
            displayName = username
        End If

        ' Check if username is taken
        If Await LocalStorageService.IsUsernameTakenAsync(username) Then
            Return New RegistrationResult(False, "Username is already taken")
        End If

        ' Create new profile
        Dim profile As New UserProfile(username, displayName, email)
        profile.PasswordHash = HashPassword(password)
        profile.LastLoginDate = DateTime.Now

        ' Save profile
        Dim saved = Await LocalStorageService.SaveProfileAsync(profile)
        If Not saved Then
            Return New RegistrationResult(False, "Failed to save profile")
        End If

        ' Auto-login after registration
        CurrentUser = profile
        Await LocalStorageService.SaveCurrentUserIdAsync(profile.UserId)
        RaiseEvent AuthStateChanged(Nothing, True)

        Return New RegistrationResult(True, "Account created successfully!", profile)
    End Function

    ''' <summary>
    ''' Attempts to log in with username and password
    ''' </summary>
    Public Shared Async Function LoginAsync(username As String, password As String) As Task(Of LoginResult)
        If String.IsNullOrWhiteSpace(username) OrElse String.IsNullOrWhiteSpace(password) Then
            Return New LoginResult(False, "Username and password are required")
        End If

        ' Find user by username
        Dim profile = Await LocalStorageService.LoadProfileByUsernameAsync(username)
        If profile Is Nothing Then
            Return New LoginResult(False, "Username not found")
        End If

        ' Verify password
        If Not VerifyPassword(password, profile.PasswordHash) Then
            Return New LoginResult(False, "Incorrect password")
        End If

        ' Update last login
        profile.LastLoginDate = DateTime.Now
        Await LocalStorageService.SaveProfileAsync(profile)

        ' Set current user
        CurrentUser = profile
        Await LocalStorageService.SaveCurrentUserIdAsync(profile.UserId)
        RaiseEvent AuthStateChanged(Nothing, True)

        Return New LoginResult(True, "Login successful!", profile)
    End Function

    ''' <summary>
    ''' Logs out the current user
    ''' </summary>
    Public Shared Async Function LogoutAsync() As Task
        CurrentUser = Nothing
        Await LocalStorageService.ClearCurrentUserAsync()
        RaiseEvent AuthStateChanged(Nothing, False)
    End Function

    ''' <summary>
    ''' Tries to restore session from stored user ID
    ''' Call this at app startup
    ''' </summary>
    Public Shared Async Function TryRestoreSessionAsync() As Task(Of Boolean)
        Try
            Dim userId = Await LocalStorageService.GetCurrentUserIdAsync()
            If String.IsNullOrEmpty(userId) Then
                Return False
            End If

            Dim profile = Await LocalStorageService.LoadProfileAsync(userId)
            If profile Is Nothing Then
                Return False
            End If

            ' Check if remember login is enabled
            If Not profile.Preferences.RememberLogin Then
                Await LocalStorageService.ClearCurrentUserAsync()
                Return False
            End If

            CurrentUser = profile
            RaiseEvent AuthStateChanged(Nothing, True)
            Return True
        Catch
            Return False
        End Try
    End Function

    ''' <summary>
    ''' Updates the current user's profile
    ''' </summary>
    Public Shared Async Function UpdateProfileAsync(displayName As String, email As String, bio As String) As Task(Of Boolean)
        If CurrentUser Is Nothing Then
            Return False
        End If

        CurrentUser.DisplayName = displayName
        CurrentUser.Email = email
        CurrentUser.Bio = bio

        Return Await LocalStorageService.SaveProfileAsync(CurrentUser)
    End Function

    ''' <summary>
    ''' Updates user preferences
    ''' </summary>
    Public Shared Async Function UpdatePreferencesAsync(preferences As UserPreferences) As Task(Of Boolean)
        If CurrentUser Is Nothing Then
            Return False
        End If

        CurrentUser.Preferences = preferences
        Return Await LocalStorageService.SaveProfileAsync(CurrentUser)
    End Function

    ''' <summary>
    ''' Changes the user's password
    ''' </summary>
    Public Shared Async Function ChangePasswordAsync(currentPassword As String, newPassword As String) As Task(Of Boolean)
        If CurrentUser Is Nothing Then
            Return False
        End If

        ' Verify current password
        If Not VerifyPassword(currentPassword, CurrentUser.PasswordHash) Then
            Return False
        End If

        ' Validate new password
        If String.IsNullOrWhiteSpace(newPassword) OrElse newPassword.Length < 4 Then
            Return False
        End If

        CurrentUser.PasswordHash = HashPassword(newPassword)
        Return Await LocalStorageService.SaveProfileAsync(CurrentUser)
    End Function

    ''' <summary>
    ''' Deletes the current user's account
    ''' </summary>
    Public Shared Async Function DeleteAccountAsync() As Task(Of Boolean)
        If CurrentUser Is Nothing Then
            Return False
        End If

        Dim userId = CurrentUser.UserId
        Await LogoutAsync()
        Return Await LocalStorageService.DeleteProfileAsync(userId)
    End Function

#Region "Password Hashing"

    ''' <summary>
    ''' Creates a simple hash of the password
    ''' Note: For production, use a proper password hashing library
    ''' </summary>
    Private Shared Function HashPassword(password As String) As String
        Using sha256 As SHA256 = SHA256.Create()
            ' Add a simple salt based on password length (basic security for local storage)
            Dim saltedPassword = $"fort.ind_{password}_uwp"
            Dim bytes = sha256.ComputeHash(Encoding.UTF8.GetBytes(saltedPassword))
            Return Convert.ToBase64String(bytes)
        End Using
    End Function

    ''' <summary>
    ''' Verifies a password against a stored hash
    ''' </summary>
    Private Shared Function VerifyPassword(password As String, storedHash As String) As Boolean
        Dim hash = HashPassword(password)
        Return hash = storedHash
    End Function

#End Region

End Class

''' <summary>
''' Result of a registration attempt
''' </summary>
Public Class RegistrationResult
    Public Property Success As Boolean
    Public Property Message As String
    Public Property Profile As UserProfile

    Public Sub New(success As Boolean, message As String, Optional profile As UserProfile = Nothing)
        Me.Success = success
        Me.Message = message
        Me.Profile = profile
    End Sub
End Class

''' <summary>
''' Result of a login attempt
''' </summary>
Public Class LoginResult
    Public Property Success As Boolean
    Public Property Message As String
    Public Property Profile As UserProfile

    Public Sub New(success As Boolean, message As String, Optional profile As UserProfile = Nothing)
        Me.Success = success
        Me.Message = message
        Me.Profile = profile
    End Sub
End Class

File: Fort.ind UWP/Services/ProfileService.cs
using System;
using System.Diagnostics;
using System.Threading.Tasks;

namespace Fort.ind_UWP
{
    public class ProfileService
    {
        public static UserProfile CurrentUser { get; set; }

        public static event EventHandler<bool> AuthStateChanged;

        public static async Task<LoginResult> LoginWithMisskeyAsync()
        {
            var result = await MisskeyAuthService.SignInAsync();
            if (!result.Success)
            {
                return new LoginResult(false, result.ErrorMessage);
            }

            await ApplySignInResultAsync(result);
            return new LoginResult(true, "Signed in!", result.Profile);
        }

        public static async Task<bool> ApplySignInResultAsync(MisskeyAuthResult result)
        {
            if (result == null || !result.Success) return false;

            CurrentUser = result.Profile;
            await LocalStorageService.SaveProfileAsync(result.Profile);
            AuthStateChanged?.Invoke(null, true);

            var displayName = string.IsNullOrWhiteSpace(result.Profile.DisplayName) ? result.Profile.Username : result.Profile.DisplayName;
            LiveTileService.SendToast("Welcome back!", $"Signed in as {displayName}");

            return true;
        }

        public static async Task LogoutAsync()
        {
            var name = CurrentUser != null ? (string.IsNullOrWhiteSpace(CurrentUser.DisplayName) ? CurrentUser.Username : CurrentUser.DisplayName) : "";
            CurrentUser = null;
            MisskeyAuthService.ClearToken();
            await LocalStorageService.ClearProfileAsync();
            AuthStateChanged?.Invoke(null, false);
            if (!string.IsNullOrEmpty(name)) LiveTileService.SendToast("Signed out", $"Goodbye, {name}!");
        }

        public static async Task ResetAppDataAsync()
        {
            MisskeyAuthService.ClearToken();
            CurrentUser = null;
            LiveTileService.ClearTile();
            LiveTileService.ClearBadge();
            await LocalStorageService.ResetAllAppDataAsync();
            AuthStateChanged?.Invoke(null, false);
        }

        public static async Task<bool> TryRestoreSessionAsync()
        {
            try
            {
                var token = await MisskeyAuthService.TryGetTokenAsync();
                if (string.IsNullOrEmpty(token))
                {
                    return false;
                }

                var cached = await LocalStorageService.LoadProfileAsync();
                if (cached == null)
                {
                    var fetched = await MisskeyAuthService.FetchCurrentUserAsync(token);
                    if (fetched == null)
                    {
                        await LogoutAsync();
                        return false;
                    }

                    CurrentUser = fetched;
                    await LocalStorageService.SaveProfileAsync(fetched);
                    AuthStateChanged?.Invoke(null, true);
                    return true;
                }

                CurrentUser = cached;
                AuthStateChanged?.Invoke(null, true);

                RefreshCurrentUserInBackground(token);

                return true;
            }
            catch (Exception ex)
            {
                Debug.WriteLine($"TryRestoreSessionAsync failed: {ex}");
                return false;
            }
        }

        private static async void RefreshCurrentUserInBackground(string token)
        {
            try
            {
                var fetched = await MisskeyAuthService.FetchCurrentUserAsync(token);
                if (fetched == null) return;

                if (!string.Equals(await MisskeyAuthService.TryGetTokenAsync(), token, StringComparison.Ordinal))
                {
                    return;
                }

                CurrentUser = fetched;
                await LocalStorageService.SaveProfileAsync(fetched);
                AuthStateChanged?.Invoke(null, true);
            }
            catch (Exception ex)
            {
                Debug.WriteLine($"RefreshCurrentUserInBackground failed: {ex.Message}");
            }
        }
    }

    public class LoginResult
    {
        public bool Success { get; set; }
        public string Message { get; set; }
        public UserProfile Profile { get; set; }

        public LoginResult(bool success, string message, UserProfile profile = null)
        {
            this.Success = success;
            this.Message = message;
            this.Profile = profile;
        }
    }
}
